### ISSUE TITLE
Create DMG background image

### ISSUE THREAD

**Issue by Daenyth** [Cockatrice/Cockatrice#2256]
DMG installer can have a background image set

https://cmake.org/cmake/help/v3.0/module/CPackDMG.html

Here's an example of a style that would work for us:

![i](https://c2.staticflickr.com/2/1049/3163487641_1400ab02b4.jpg)

**Comment by ZeldaZach** [comment]
@tooomm Wanna check this out?

**Comment by SlightlyCircuitous** [comment]
This seemed like fun so did some work on it. I made a .dmg manually in macOS 10.13 (only mac environment I have access to) as a mock up that seems to work well and can install both Cockatrice and Oracle by dragging. 

This is the layout of my mock up. I tried a few different colors and I'm open to other suggestions. As you can see, the icon label text doesn't show up as well on some of them. I think I like the blue one the best currently. The font is Noto Sans, which matches the website, but I'm happy to change it if there's a different application font or anything.

![magentatrice](https://user-images.githubusercontent.com/71394296/208986138-239ede1f-eb26-429b-9f0b-41492d59e167.png)

![bluatrice](https://user-images.githubusercontent.com/71394296/208986619-ce7682f4-1d7e-42d1-b90d-61afa4268091.png)

![purpatrice](https://user-images.githubusercontent.com/71394296/208986660-692a0af0-fd09-43d6-808a-11369ca37ac2.png)

My next step is to get the building process automated and integrated, but I wanted to check in about the goals first. Do we just need a background picture and a .DS_store file to facilitate building carbon copies of a predefined layout with cmake Drag'n'Drop or do we want a full AppleScript like in the above cmake announcement that can be modified to build any layout .dmg from scratch? From reading the above links, it seems like we can do either, just depends on how much flexibility/future proofing we want in the build. Lemme know either way and I'll keep playing around with it.

**Comment by ebbit1q** [comment]
I know there have been some changes with installing on later mac versions but I wouldn't know if this is affected, as for colors they all seem to have terrible fringing around the smaller text and it's the least bad with blue?
as for the method of automating it, ease of maintainability is most important, if we need an apple os ourselves to configure it that would make adjustments less easy.

**Comment by SlightlyCircuitous** [comment]
I'll take another pass at the colors. I am not sure why that white fringe happens so it's been a little trial and error. Maybe it doesn't like the gradient.

As far as automating it goes, I don't think we can get around using macOS to make changes. Apple has set it up so you basically need to use Finder to build these because that's the only way to get the DS_store file.

It seems like if I just give you a DS_store file and a background image, you can build it on any OS using cmake, but it will be arranged exactly how it is now, so you'd need macOS to make adjustments and generate a new DS_store file if you ever wanted to change it. Same thing with the AppleScript, as that just automates the manual process of arranging the icons and getting the DS_store file, so it also has to run on macOS and talk to Finder.

**Comment by SlightlyCircuitous** [comment]
I learned that you can pretty easily make a .dmg with a background image on Linux using genisoimage. Built one using the .app files for Cockatrice/Oracle 2.7.5 and I was able to use it to drag to install both in my macOS 10.13 test environment. I was hoping to use CPack DragNDrop to make it since I know we use a lot of cmake, but DraganDrop is apparently exclusive to macOS. 

Using genisoimage still needs a .DS_Store file created on macOS though or it makes a .dmg like the one we already package (white background, no icon arrangement). So, you'd be reliant on macOS for that, but I think you could pack whatever you wanted on Linux as long as you didn't need to change the background image or icon position. There seems to be at least one node.js tool that makes DS_Store files on Linux, but I haven't looked into it at all yet. I don't know if that's something worth looking at if we want to be completely macOS independent.

I still don't know what's going on with the white fringe on the text. It's clearly a general issue with using a background image since it doesn't happen if you fill with a solid color, but I haven't been able to find a fix so far.

**Comment by SlightlyCircuitous** [comment]
After some testing on macOS 12, the white outline seems to be a visual bug on 10.13 that doesn't appear on more modern macOS versions. So, it should be fine to ignore it as we don't even support that version (or 10.14) anymore.

**Comment by SlightlyCircuitous** [comment]
I have been continuing to fiddle around with this in my spare time and as far as I can tell, it really isn't feasible to build a dmg with a background image without using macOS at some point. While I can build one with genisoimage, that requires a ds_store file that was created by Finder and, in my testing, it also seems like it specifically requires a ds_store file created by the version of macOS that you want to build for (my build using ds_store from macOS 10.13 didn't display a background on macOS 11, it just opened like a normal folder). I also assume that we don't want to use an arbitrary file that I created as a central part of the build so someone would have to make a new mock-up and generate a new ds_store in some trusted environment.

While I did find [a python library that can read and edit ds_store files](https://github.com/dmgbuild/ds_store), its very hard to know what exactly to change to get things to work right given how poorly documented these files are (Apple doesn't publish any info, any docs are reverse engineered guesswork). I'm really not even sure where to begin there. Really, it seems like that tool is not supposed to be run in isolation and is instead supposed to be used in combination with the rest of the [dmgbuild library](https://github.com/dmgbuild) to automate building a dmg with a background. However, dmgbuild has to run on macOS so it has access to the macOS-specific tools and information needed to fiddle around with the ds_store so it works right. 

In summary, both automated tools I looked into (CPack Drag'n'Drop and dmgbuild) need to run on macOS to work at all. While using genisoimage *can* build a dmg on Linux, it is limited by needing to use an already created ds_store file, which requires macOS and also seems to bring with it some versioning issues.

**Comment by ebbit1q** [comment]
this intercompatibility issue between mac versions seems troublesome, I was under the impression that software images created on older versions were forwards compatible. we also don't support macos 10.13 anymore. I'm not sure what you are stuck on but we are building on mac, building on mac is not an issue as we have access to mac runners in github actions.

**Comment by SlightlyCircuitous** [comment]
I was under the impression that you wanted a solution that did not involve using macOS, although looking back in the thread, I guess you just said it would be 'less easy' to need to use macOS, not impossible.

**Comment by SlightlyCircuitous** [comment]
Using GitHub actions and some AppleScript code linked in tooomm's comment, I got CPack Drag'n'Drop working in my fork and made a DMG installer that looks like this:

<img width="995" alt="Screen Shot 2023-04-15 at 9 01 22 PM" src="https://user-images.githubusercontent.com/71394296/232260655-0ae35e3c-6b85-444e-aedf-19f0e66dd539.png">

This would require adding an AppleScript (.script) file and a background image (.tif) to the repo as well as modifying CMakeLists so Cpack knows where to find those things.

The icon size and position and the size of the window are set by the AppleScript, so they would be readily editable. I made the background (with the arrows and the text) in GIMP, but I could post the .xcf file somewhere convenient so that it could be readily editable as well. 

By default, CPack wants to draw icons for Cockatrice, Oracle, Servatrice, and dbconverter. I have drawn dbconverter outside of the bounds of the window since I figured most people wouldn't want or need to install it, but I could work it into the design if we wanted it included. I have included Servatrice in the mock up since some people might want it, but I tried my best to make it look optional. Open to suggestions on layout, color, or what text should be included.

I used the Big Sur build for testing since I have access to a computer running Big Sur, but, in theory, this was also built for 10.15. I haven't looked at that one since 10.15 is no longer supported by Apple as of last November and I assumed it would be removed soon, but I can make sure it's OK if we want to continue to support 10.15.

I'm still trying to figure out if it's possible to make the icon text bigger and if there is a better way to hide an icon than drawing it out of frame. Also, in typing this up, I noticed that I spelled Servatrice wrong so I'll fix that.

**Comment by ebbit1q** [comment]
looks fine, please make a pr so I can take a look

**Comment by SlightlyCircuitous** [comment]
will do
